components/allowances/dashboard/cells/LastCancelledCell.tsx

```diff
@@ -0,0 +1,56 @@
+import Href from 'components/common/Href';
+import WithHoverTooltip from 'components/common/WithHoverTooltip';
+import { ADDRESS_ZERO_PADDED, ADDRESS_ZERO_PADDED_1 } from 'lib/constants';
+import { useAddressEvents } from 'lib/hooks/page-context/AddressPageContext';
+import type { AddressEvents, AllowanceData } from 'lib/interfaces';
+import { getChainExplorerUrl } from 'lib/utils/chains';
+import { SECOND, formatDateNormalised } from 'lib/utils/time';
+import useTranslation from 'next-translate/useTranslation';
+import TimeAgo from 'timeago-react';
+
+interface Props {
+  allowance: AllowanceData;
+}
+
+// We export this function so it can be reused in other places such as the 'PermitsEntry.tsx' file.
+export function filterLastCancelled(
+  events: AddressEvents,
+  token: AllowanceData,
+): {
+  time: string;
+  alreadyCancelled: boolean;
+} {
+  const filteredEvents = events.approval.filter(
+    (event) =>
+      event.address === token.contract.address &&
+      (event.topics[2] === ADDRESS_ZERO_PADDED_1 ||
+        event.topics[2] === ADDRESS_ZERO_PADDED ||
+        (event.timestamp > 0 && event.topics[2] !== ADDRESS_ZERO_PADDED)) &&
+      (event.data === ADDRESS_ZERO_PADDED || event.data === '0x'),
+  );
+  // If the Log list contain at least an element, then we are sure to retrieve the lastCancelled time
+  if (filteredEvents?.length > 0) {
+    return { time: formatDateNormalised(new Date(filteredEvents[0]?.timestamp * SECOND)), alreadyCancelled: true };
```

Comment:
Most events we receive (on Ethereum) don't contain a timestamp, only a block number. To get the timestamp for that block, you need to import `blocksDB` from `lib/databases/blocks` and call `await blocksDB.getBlockTimestamp(publicClient, blockNumber)`.

Doing that turns this into an async operation. That means that we should probably do this operation in `PermitsPanel.tsx` in the spot where we filter our tokens so that we don't need to inroduce more async functions (https://github.com/RevokeCash/revoke.cash/blob/master/components/signatures/permit/PermitsPanel.tsx#L34).